packages/dds/map/src/test/mapOracle.ts

```diff
@@ -0,0 +1,71 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { strict as assert } from "node:assert";
+
+import type { ISharedMap, IValueChanged } from "../interfaces.js";
+
+/**
+ * Simple oracle for ISharedMap that mirrors the map state.
+ * @internal
+ */
+export class SharedMapOracle {
+	private readonly oracle = new Map<string, unknown>();
+
+	public constructor(private readonly fuzzMap: ISharedMap) {
+		// Snapshot current state
+		for (const [k, v] of fuzzMap.entries()) {
+			this.oracle.set(k, v);
+		}
+		// Subscribe
+		this.fuzzMap.on("valueChanged", this.onValueChanged);
+		this.fuzzMap.on("clear", this.onClear);
+	}
+
+	private readonly onValueChanged = (change: IValueChanged): void => {
+		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+		const { key, previousValue } = change;
+
+		assert.strictEqual(
+			previousValue,
+			this.oracle.get(key),
+			`Mismatch on previous value for key="${key}"`,
+		);
+
+		if (this.fuzzMap.has(key)) {
+			// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+			const newVal = this.fuzzMap.get(key);
+			this.oracle.set(key, newVal); // key exists, even if value is undefined
+		} else {
+			this.oracle.delete(key); // key was deleted
+		}
+	};
+
+	/**
+	 * Note: Simply clearing the oracle can leave it out of sync with the DDS, since other clients may re-populate entries immediately after a `clear`. To keep the oracle consistent, we rebuild it from the current state of the fuzzMap whenever a `clear` event is observed.
```

Comment:
this feels like a bug to me. do you see the same behavior for local and remote clears? i could see this happening if there is a remote clear, and there are pending local changes? is that the case where it happens? local clears should always clear all values. can we make the oracle better represent this, and file a bug for case where clear is fired by the map isn't cleared

Reply:
i'm not quite following. i think there is a bug in the existing map code, specifically, that in some case we fire a clear event, but the map is not cleared. Specifically, in the clear event handle if the map is read, there should be no keys in the map. 

The reason i think this is a bug, is that it makes it no very useful for consumer, as they cannot de[end on the clear event, they need to fallback to reading keys like you do here.

can you file a bug for this problem, and call it out in a comment in the code. no need to fix immediately, i just want the issue captured.